File: android_native/docx/src/main/kotlin/com/afluffywaffle/layuv/docx/JsonWriter.kt
package com.afluffywaffle.layuv.docx

/**
 * Compact JSON serializer for `leamh/annotations.json` matching Dart's
 * `jsonEncode` escaping (escape `"` `\` and control chars as Dart does; leave
 * `/` and non-ASCII raw — unlike org.json, which escapes `/`). Output is valid
 * JSON the Dart store reads back identically. Byte-identity with Dart isn't
 * required (annotations.json is compared semantically), only Dart-readability.
 */
internal object JsonWriter {
    fun encode(value: Any?): String = StringBuilder().also { write(it, value) }.toString()

    private fun write(sb: StringBuilder, v: Any?) {
        when (v) {
            null -> sb.append("null")
            is String -> writeString(sb, v)
            is Boolean -> sb.append(v.toString())
            is Int, is Long -> sb.append(v.toString())
            is Double -> sb.append(encodeDouble(v))
            is Map<*, *> -> {
                sb.append('{')
                var first = true
                for ((k, vv) in v) {
                    if (!first) sb.append(',')
                    first = false
                    writeString(sb, k.toString())
                    sb.append(':')
                    write(sb, vv)
                }
                sb.append('}')
            }
            is List<*> -> {
                sb.append('[')
                var first = true
                for (e in v) {
                    if (!first) sb.append(',')
                    first = false
                    write(sb, e)
                }
                sb.append(']')
            }
            else -> writeString(sb, v.toString())
        }
    }

    private fun writeString(sb: StringBuilder, s: String) {
        sb.append('"')
        for (c in s) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                else -> if (c < ' ') sb.append("\\u%04x".format(c.code)) else sb.append(c)
            }
        }
        sb.append('"')
    }

    private fun encodeDouble(d: Double): String =
        if (!d.isInfinite() && !d.isNaN() && d == d.toLong().toDouble()) "${d.toLong()}.0" else d.toString()
}

File: macos_native/Packages/LeamhDocx/Sources/LeamhDocx/JsonWriter.swift
import Foundation

/// Compact JSON serializer for leamh/annotations.json matching Dart's jsonEncode escaping.
/// Escapes " \ and control chars; leaves / and non-ASCII raw (unlike JSONSerialization which escapes /).
/// Mirrors JsonWriter.kt.
enum JsonWriter {

    static func encode(_ value: Any?) -> String {
        var sb = ""
        write(&sb, value)
        return sb
    }

    private static func write(_ sb: inout String, _ v: Any?) {
        switch v {
        case nil, is NSNull:
            sb += "null"
        case let s as String:
            writeString(&sb, s)
        case let b as Bool:
            sb += b ? "true" : "false"
        case let n as NSNumber:
            // Distinguish Bool from numeric (Bool is NSNumber in ObjC bridge)
            if n === kCFBooleanTrue as AnyObject || n === kCFBooleanFalse as AnyObject {
                sb += n.boolValue ? "true" : "false"
            } else {
                sb += encodeNumber(n)
            }
        case let d as [String: Any?]:
            sb += "{"
            var first = true
            for (k, vv) in d {
                if !first { sb += "," }
                first = false
                writeString(&sb, k)
                sb += ":"
                write(&sb, vv)
            }
            sb += "}"
        case let arr as [Any?]:
            sb += "["
            var first = true
            for e in arr {
                if !first { sb += "," }
                first = false
                write(&sb, e)
            }
            sb += "]"
        default:
            writeString(&sb, "\(v!)")
        }
    }

    private static func writeString(_ sb: inout String, _ s: String) {
        sb += "\""
        for c in s.unicodeScalars {
            switch c {
            case "\"": sb += "\\\""
            case "\\": sb += "\\\\"
            case "\n": sb += "\\n"
            case "\r": sb += "\\r"
            case "\t": sb += "\\t"
            case "\u{08}": sb += "\\b"
            case "\u{0C}": sb += "\\f"
            case _ where c.value < 0x20:
                sb += String(format: "\\u%04x", c.value)
            default:
                sb += String(c)
            }
        }
        sb += "\""
    }

    private static func encodeNumber(_ n: NSNumber) -> String {
        let d = n.doubleValue
        if !d.isInfinite && !d.isNaN && d == Double(Int64(exactly: n.doubleValue) ?? 0) {
            if let i = Int64(exactly: n.doubleValue) { return "\(i).0" }
        }
        // Integer types
        let objCType = String(cString: n.objCType)
        switch objCType {
        case "i", "s", "l", "q", "I", "S", "L", "Q":
            return "\(n.int64Value)"
        default:
            return "\(d)"
        }
    }
}
